### ISSUE TITLE
Get disparity maps that aids in 3D Reconstruction

### ISSUE THREAD

**Issue by abhileshborode** [jiuerbujie/omnidirectional_calibration#6]
hey,

The aim of my project is to generate disparity maps from a stereo fisheye camera. I have calibrated my camera using the omnidirectional module. I am using the stereoreconstruct functions to generate the  rectified images and then I use the post-filtering module by OpenCV for better disparity maps by filling the holes in the disparity map. The problem I am facing is that both the rectified images differ in size a little bit. (Using RECTIFY_LONGI) Due to which I am getting non zero disparity values at the portions which do not overlap in the left and right images. For example left image
![d2](https://user-images.githubusercontent.com/33388291/35316863-15d7c914-0091-11e8-9d0a-dd0961cdbd0d.png)
right image
![d1](https://user-images.githubusercontent.com/33388291/35316852-0d2aabce-0091-11e8-8be4-872d9f794839.png)
disparity image
![dd](https://user-images.githubusercontent.com/33388291/35316906-4bf912f0-0091-11e8-89c1-2a67c57abd29.png)

As you can see there are some non zero values in the disparity image like the portion outside the curved boundary of the cropped rectified images. This is causing me problems while performing 3D reconstruction of the disparity map as it also considers the portion outside the actual image. Generating the rectified images of the same cropping size for left and right images will do the trick for me.  

Please let me know how I can proceed with it. Or if you could tell me how to fill the holes in the map created by stereoreconstruction function to get a smooth disparity map, that would be fine too.

Thanks

**Comment by jiuerbujie** [comment]
Hi, 
   I have a trick for you. Maybe you can use your camera to take a photo for a white paper or something else which is easy to segment out. Then you can get the boundary of field of view.

**Comment by abhileshborode** [comment]
hey ,
I apologize I don't understand your answer fully could you please elaborate?

**Comment by jiuerbujie** [comment]
I mean, you can take a photo of something bright, then you get a picture which has only weight and black regions, and the black ones are all out side of field of view (" the portion outside the curved boundary" as you said). You can use a simple threshold method to get black regions. The disparity in such black regions can manually set to be 0.
Note this is just a trick method.

**Comment by abhileshborode** [comment]
oh I get it now. Also could you recommend any technique to fill out the holes present in the disparity map.

**Comment by jiuerbujie** [comment]
@abhileshborode The StereoSGBM class of OpenCV has two parameters (p1 and p2) to control the smoothness of disparity, you can first try adjust them. If this does not work, may you need to move to image 
 inpainting  techniques.

**Comment by abhileshborode** [comment]
Got it! My last question is about pointClouds mentioned in the tutorials of your omnidirectional repository. Could you tell me how to display that pointCloud image? 

Also when you performed the 3D reconstruction for your disparity maps do you remember how accurate were your depth values for a corresponding point in real word coordinates. As my checkerboard size was  0.025 meters and the range of depth values i am getting is starting from 3-4 metres for the closed object. However that objects actual depth in real world is around 0.4 meters. So its completely off scale. Although the depth values are relatively consistent for close and far away objects. That is far away objects have more depth than closer ones ( which makes sense). Could you please suggest me how to improve my depth accuracy to the scale ?

**Comment by jiuerbujie** [comment]
I use CloudCompare to display point cloud.
As https://github.com/jiuerbujie/omnidirectional_calibration/blob/26067fabe3e58260b39b6f64139341a04b1261d2/src/omnidir.cpp#L1481, I filter out points that has disparity less than 15, maybe you need to adjust this parameter.

**Comment by abhileshborode** [comment]
I had a doubt regarding the stereo reconstruction, I am using a fisheye camera and I used the omnidirectional module along with the math involved in reconstruction instead of the fisheye module for the same. Is that why the 

problem is created? 

Also is there any function by which i can distort my disparity image points(project the points) to the original field of view of the camera.? For example undo the image rectification for the input images. I want to do the same for the generated disparity image. As shown in the image below

![capture](https://user-images.githubusercontent.com/33388291/35609925-a29f9814-061c-11e8-9b41-2953b44ea2c7.PNG)

**Comment by jiuerbujie** [comment]
Sorry for the delayed reply. The omnidirectional camera model in this project is suitable for fisheye camera, so I do not think this is a problem. 

I will recommend you first apply RECTIFY_PERSPECTIVE for the reconstruction and get a reliable scale and check if the scale of RECTIFY_LONGLATI is consistent with it. If they are not the same, please let me know, there may be some bugs in the reconstruction code.

It is possible to undo the image rectification, but there is no such interfaces for that. You may have to implement the inverse of omnidir::initUndistortRectifyMap function. But I can't see its necessity.

**Comment by abhileshborode** [comment]
("You may have to implement the inverse of omnidir::initUndistortRectifyMap function."), yes i understood that i will have to take the inverse.  But without an interface I dont know how to undo the image rectification. I though of using projectpoints but that would just give me the distorted multiple points instead of an image. if you could tell me how to do the inverse?
    
 ("But I can't see its necessity")
So i wanted to project the resulting disparity map image onto the original cameras FOV like the image  atttched in my previous reply. Which makes it easier to look at especially while you are using your robot for navigation purposes

**Comment by jiuerbujie** [comment]
emm, I see. The easy and tricky way is to consider the rectified images (what you used for stereo matching) as your robot's view. It should be ok because some stereo cameras just also only output the rectified images. 

For the hard  way, there is no existing API you can use. You need to implement a ```distortImage``` function. You may refers to ```undistortImage``` function and first understand the math of camera models.

**Comment by abhileshborode** [comment]
oh I see. I thought there must be an in built function since the people in the image attached got it. Never mind thanks for answering all my questions.
